Failure looks related

```
  error: configuration error in .github
    |
    = help: check your configuration file for syntax errors
    = help: see: https://docs.zizmor.sh/configuration/
  
  Caused by:
      0: configuration error in .github
      1: invalid configuration syntax
      2: rules.rules: unknown field `unpinned-uses`, expected one of `disable`, `ignore`, `config`, `remap` at line 14 column 5
  Error: Process completed with exit code 1.
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -60,6 +60,25 @@ jobs:
         with:
           base_ref: ${{ github.event.pull_request.base.ref }}
 
+  zizmor:
+    permissions:
+      contents: read
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    steps:
+      - uses: actions/checkout@v7.0.1
+        with:
+          persist-credentials: false
+      - name: Run zizmor
+        uses: zizmorcore/zizmor-action@3dc1ecc9bcb9e94e9b2c709687979e1298497054 # v0.6.2
+        with:
+          inputs: .github
+          # Version of the zizmor tool to install and run
+          version: 1.29.0
+          # Fail the job and annotate the PR instead of reporting to GitHub Advanced Security
+          advanced-security: false
+          annotations: true
+
   path-filters:
     runs-on: ubuntu-latest
     outputs:
@@ -1214,6 +1233,7 @@ jobs:
       - test
       - test-jdbc-compatibility
       - test-other-modules
+      - zizmor
     steps:
       - name: "Check results"
         run: |
@@ -1231,3 +1251,4 @@ jobs:
           echo '${{ needs.test.result }}' | grep -xE 'success|skipped' || { echo 'Job "test" failed' >&2; exit 1; }
           echo '${{ needs.test-jdbc-compatibility.result }}' | grep -xE 'success|skipped' || { echo 'Job "test-jdbc-compatibility" failed' >&2; exit 1; }
           echo '${{ needs.test-other-modules.result }}' | grep -xE 'success|skipped' || { echo 'Job "test-other-modules" failed' >&2; exit 1; }
+          echo '${{ needs.zizmor.result }}' | grep -xE 'success|skipped' || { echo 'Job "zizmor" failed' >&2; exit 1; }

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -6,6 +6,10 @@ on:
   repository_dispatch:
     types: [test-with-secrets-command]
 
+# Default for all jobs; jobs that need more declare their own permissions block, which fully replaces this one.
+permissions:
+  contents: read
+
 defaults:
   run:
     shell: bash --noprofile --norc -euo pipefail {0}
@@ -80,6 +84,8 @@ jobs:
           annotations: true
 
   path-filters:
+    permissions:
+      pull-requests: read # to list changed files of the PR
     runs-on: ubuntu-latest
     outputs:
       ci: ${{ steps.filter.outputs.ci }}
@@ -98,6 +104,9 @@ jobs:
             non_docs: '!docs/**'
 
   maven-checks:
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure
+      contents: read
     runs-on: ubuntu-latest
     name: maven-checks ${{ matrix.java-version }}
     strategy:
@@ -143,6 +152,9 @@ jobs:
         run: .github/bin/cancel-merge-queue-workflow.sh "maven-checks ${{ matrix.java-version }}"
 
   artifact-checks:
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure
+      contents: read
     needs: path-filters
     if: github.event_name != 'pull_request' || needs.path-filters.outputs.non_docs == 'true'
     runs-on: ubuntu-latest
@@ -242,6 +254,9 @@ jobs:
           base_ref: ${{ github.event.pull_request.base.ref }}
 
   error-prone-checks:
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure
+      contents: read
     needs: path-filters
     if: github.event_name != 'pull_request' || needs.path-filters.outputs.non_docs == 'true'
     runs-on: ubuntu-latest
@@ -278,6 +293,9 @@ jobs:
         run: .github/bin/cancel-merge-queue-workflow.sh "error-prone-checks"
 
   test-jdbc-compatibility:
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure
+      contents: read
     needs: path-filters
     if: github.event_name != 'pull_request' || needs.path-filters.outputs.non_docs == 'true'
     runs-on: ubuntu-latest
@@ -326,6 +344,9 @@ jobs:
         run: .github/bin/cancel-merge-queue-workflow.sh "test-jdbc-compatibility"
 
   test-other-modules:
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure
+      contents: read
     needs: path-filters
     if: github.event_name != 'pull_request' || needs.path-filters.outputs.non_docs == 'true'
     runs-on: ubuntu-latest
@@ -420,6 +441,12 @@ jobs:
         run: .github/bin/cancel-merge-queue-workflow.sh "test-other-modules"
 
   build-test-matrix:
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure, read run info in update-check
+      checks: write # to update the PR check for test-with-secrets runs (update-check)
+      contents: read
+      issues: write # to comment on the PR for test-with-secrets runs (update-check)
+      pull-requests: read # to read the PR head SHA in update-check
     needs: path-filters
     if: github.event_name != 'pull_request' || needs.path-filters.outputs.non_docs == 'true'
     runs-on: ubuntu-latest
@@ -545,6 +572,13 @@ jobs:
         run: .github/bin/cancel-merge-queue-workflow.sh "build-test-matrix"
 
   test:
+    name: ${{ matrix.name != '' && matrix.name || format('test ({0}, {1})', matrix.modules, matrix.profile) }}
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure, read run info in update-check
+      checks: write # to update the PR check for test-with-secrets runs (update-check)
+      contents: read
+      issues: write # to comment on the PR for test-with-secrets runs (update-check)
+      pull-requests: read # to read the PR head SHA in update-check
     runs-on: 'ubuntu-latest'
     needs: build-test-matrix
     if: needs.build-test-matrix.outputs.matrix != '{}'
@@ -875,6 +909,9 @@ jobs:
         run: .github/bin/cancel-merge-queue-workflow.sh "test (${{ matrix.modules }}, ${{ matrix.profile }}, ${{ matrix.jdk }})"
 
   build-pt:
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure
+      contents: read
     needs: path-filters
     if: github.event_name != 'pull_request' || needs.path-filters.outputs.non_docs == 'true'
     runs-on: ubuntu-latest
@@ -957,6 +994,12 @@ jobs:
         run: .github/bin/cancel-merge-queue-workflow.sh "build-pt"
 
   pt:
+    permissions:
+      actions: write # to dispatch cancel-merge-queue-workflow on failure, read run info in update-check
+      checks: write # to update the PR check for test-with-secrets runs (update-check)
+      contents: read
+      issues: write # to comment on the PR for test-with-secrets runs (update-check)
+      pull-requests: read # to read the PR head SHA in update-check
     runs-on: 'ubuntu-latest'
     name: pt (${{ matrix.bucket }})
     if: needs.build-pt.outputs.matrix != '{}'

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -93,6 +93,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits, as the build result depends on `git describe` equivalent
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&
@@ -131,6 +132,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits, as the build result depends on `git describe` equivalent
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&
@@ -174,6 +176,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits to be able to determine merge base
+          persist-credentials: false
       - name: Block illegal commits
         uses: trinodb/github-actions/block-commits@c2991972560c5219d9ae5fb68c0c9d687ffcdd10
         with:
@@ -208,6 +211,7 @@ jobs:
         if: matrix.commit != ''
         with:
           fetch-depth: 0 # checkout all commits to be able to determine merge base
+          persist-credentials: false
           ref: ${{ matrix.commit }}
       # This composite job must be entirely standalone, and checked out from the correct commit before being executed.
       # It can't accept any parameters defined in this workflow, because the values of those parameters would always be taken from
@@ -227,6 +231,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits to be able to determine merge base for GIB
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&
@@ -262,6 +267,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout tags so version in Manifest is set properly
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&
@@ -309,6 +315,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits to be able to determine merge base for GIB
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&
@@ -403,6 +410,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits to be able to determine merge base for GIB
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&
@@ -529,6 +537,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits to be able to determine merge base for GIB
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&
@@ -856,6 +865,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits, as the build result depends on `git describe` equivalent
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&
@@ -940,6 +950,7 @@ jobs:
       - uses: actions/checkout@v7.0.1
         with:
           fetch-depth: 0 # checkout all commits, as the build result depends on `git describe` equivalent
+          persist-credentials: false
           ref: |
             ${{ github.event_name == 'repository_dispatch' &&
                 github.event.client_payload.pull_request.head.sha == github.event.client_payload.slash_command.args.named.sha &&

--- a/.github/workflows/cleanup.yml
+++ b/.github/workflows/cleanup.yml
@@ -6,6 +6,9 @@ on:
     types:
       - closed
 
+permissions:
+  actions: write # to cancel workflow runs of closed PRs
+
 # NOTE: Do not check out/execute code from the PR since a read-write token is present due to pull_request_target
 jobs:
   cancel:

--- a/.github/workflows/milestone.yml
+++ b/.github/workflows/milestone.yml
@@ -15,6 +15,8 @@ jobs:
     steps:
       - name: Checkout code
         uses: actions/checkout@v7.0.1
+        with:
+          persist-credentials: false
       - uses: ./.github/actions/setup
         timeout-minutes: 15
       - name: Get milestone from pom.xml

--- a/.github/workflows/milestone.yml
+++ b/.github/workflows/milestone.yml
@@ -4,6 +4,11 @@ on:
   push:
     branches: [master]
 
+permissions:
+  contents: read
+  issues: write # to create milestones and set them on merged PRs
+  pull-requests: write
+
 defaults:
   run:
     shell: bash --noprofile --norc -euo pipefail {0}

--- a/.github/workflows/ok-to-test.yml
+++ b/.github/workflows/ok-to-test.yml
@@ -5,6 +5,10 @@ on:
   issue_comment:
     types: [created]
 
+permissions:
+  issues: write # to add a reaction to the slash command comment
+  pull-requests: write
+
 jobs:
   test-with-secrets:
     runs-on: ubuntu-latest

--- a/.github/zizmor.yml
+++ b/.github/zizmor.yml
@@ -1,0 +1,16 @@
+# Configuration for zizmor, the GitHub Actions security scanner: https://zizmor.sh/
+# Rules disabled here are off for the whole repository; findings accepted case by case
+# are instead suppressed inline with "# zizmor: ignore[rule-name]" comments in the workflows.
+rules:
+  # Flags every ${{ }} expression interpolated into a run: script, regardless of whether the
+  # value is attacker-controlled. The interpolations used here read trusted contexts only
+  # (event name, matrix values, secrets), so this audit produces only false positives.
+  template-injection:
+    disable: true
+  # Requires every action to be pinned to a commit SHA. Third-party actions are SHA-pinned,
+  # but actions authored by GitHub itself (actions/*) are trusted at version tags and kept
+  # current by Dependabot, which this rule's default policy does not allow.
+  unpinned-uses:
+    config:
+      policies:
+        actions/*: ref-pin

